src/synth/fluid_voice.c

```diff
@@ -586,6 +604,17 @@ fluid_voice_calculate_runtime_synthesis_parameters(fluid_voice_t* voice)
     fluid_voice_update_param(voice, list_of_generators_to_initialize[n]);
   }
 
+  /* Start portamento if enabled */
+  {	/* fromkey note comes from "GetFromKeyPortamentoLegato()" detector.
+	When fromkey is set to ValidNote , portamento is started */
+	  /* Return fromkey portamento */
+	  int fromkey = voice->channel->synth->fromkey_portamento;
```

Comment:
Please think of a better solution to retrieve `fromkey_portamento` here and elsewhere. The synth and therefore also `fromkey_portamento` is mutex protected, which you are ignoring. Why are you not storing this key inside `fluid_channel_t`?

Reply:
>The synth and therefore also fromkey_portamento is mutex protected, which you are ignoring.

I'm aware of the mutex, please which issue did you see on this subject ?.

>Why are you not storing this key inside fluid_channel_t ?

Simply because on each `noteOn` (regardless of the channel), this `key` is always updated by `get_fromkey_portamento_legato()`  https://github.com/FluidSynth/fluidsynth/blob/67ff72659f4e83a0e0e24912e919071ae692f590/src/synth/fluid_synth_mono.c#L475

and then this `key` is read in the context of this `noteOn` regardless of the channel concerned, so to avoid waste of memory i have chosen `fluid_synt_t` rather than  `fluid_channel_t`.

Reply:
I just dont like accessing anything via this indirection `voice->channel->synth`, yes it's done like that at several places in `fluid_voice.c`... However looking at it again I think when accessing `fromkey_portamento` that way the synth API should be locked anyway, so never mind. It's probably fine.

Reply:
I must admit I don't understand why you store fromkey_portamento on the synth instance in the first place. Couldn't you just return it in a return parameter of the get_fromkey_portamento_legato function?

Reply:
Ok, having read more of the code now, I think I understand: you need it somewhere that fluid_voice_start can reach. And passing it directly would mean changing a lot of public function signatures. Still, it seems very strange to have get_fromkey_portamento_legato return the legato fromkey directly and the portamento fromkey indirectly in such a far away place. But I don't have a good suggestion on how to do it otherwise...

Reply:
>derselbst : I think when accessing fromkey_portamento that way the synth API should be locked anyway, so never mind. It's probably fine.

On fluid_synth_noteon() , between the last update of `fromkey_portamento`  (made by get_fromkey_portamento_legato()) to the use of this `fromkey_portamento` (in  fluid_voice_start()), effectively there is a call to `FLUID_API_ENTRY_CHAN` call (in  fluid_synth_alloc_voice_LOCAL()) but the mutex involved is recursive, so it doesn't lock the current task (but should lock another task calling fluid_synth_noteon() API). In short , On noteOn, PolyMono calls fluid_preset_noteon() like on noteOn without PolyMono. I don't see a potential deadlock that wasn't already there before.

>mawe42 : And passing it directly would mean changing a lot of public function signatures.

Yes, unfortunately.

>Still, it seems very strange to have get_fromkey_portamento_legato return the legato fromkey directly and the portamento fromkey indirectly in such a far away place. But I don't have a good suggestion on how to do it otherwise..

Me too, i regret to be forced in this situation and don't have a good  suggestion on how to do it otherwise.

For information,note the same regrettable situation (not related to polymono) for `storeid` which is also a "far away parameter" for fluid_voice_init().
https://github.com/FluidSynth/fluidsynth/blob/67ff72659f4e83a0e0e24912e919071ae692f590/src/synth/fluid_synth.c#L4541-L4542.

Reply:
> I don't see a potential deadlock that wasn't already there before.

No no, there is no deadlock. I just wanted to be sure this way of accessing `fromkey_portamento` is threadsafe. And since there is a call to `FLUID_API_ENTRY_CHAN` it's fine.